### PR TITLE
[SPARK-6806] [SparkR] [Docs] Fill in SparkR examples in programming guide

### PR DESCRIPTION
**Pull request by davies**
sqlCtx -> sqlContext

You can check the docs by:

```
$ cd docs
$ SKIP_SCALADOC=1 jekyll serve
```

cc @shivaram

### CODE REVIEW

**Comment by shivaram** [comment]
@cafreeman -- If you get a chance, could you take a look at this too ?

File: docs/programming-guide.md [hunk 1]
```diff
@@ -246,6 +279,23 @@ your notebook before you start to try Spark from the IPython notebook.
...
 </div>
 
+<div data-lang="r"  markdown="1">
+
+In the SparkR shell, a special interpreter-aware SparkContext is already created for you, in the
+variable called `sc`. Making your own SparkContext will not work. You can set which master the
+context connects to using the `--master` argument. You can also add dependencies
+(e.g. Spark Packages) to your shell session by supplying a comma-separated list of maven coordinates
+to the `--packages` argument. Any additional repositories where dependencies might exist (e.g. SonaType)
+can be passed to the `--repositories` argument. For example, to run `bin/pyspark` on exactly four cores, use:
```
Comment (cafreeman) [inline docs/programming-guide.md]: This should refer to SparkR instead of PySpark.

File: docs/programming-guide.md [hunk 2]
```diff
@@ -477,8 +541,28 @@ the [Converter examples]({{site.SPARK_GITHUB_URL}}/tree/master/examples/src/main
...
+
+{% highlight r %}
+distFile <- textFile(sc, "data.txt")
+{% endhighlight %}
+
+Once created, `distFile` can be acted on by dataset operations. For example, we can add up the sizes of all the lines using the `map` and `reduce` operations as follows: `reduce(map(distFile, length), function(a, b) {a + b})`.
+
+Some notes on reading files with Spark:
+
+* If using a path on the local filesystem, the file must also be accessible at the same path on worker nodes. Either copy the file to all workers or use a network-mounted shared file system.
```
Comment (cafreeman) [inline docs/programming-guide.md]: Might be worth linking to either the EC2 article on the SparkR wiki or the Spark doc about distributing a file across a cluster (the `copy-dir` stuff):

https://github.com/amplab-extras/SparkR-pkg/wiki/SparkR-on-EC2

https://spark.apache.org/docs/latest/ec2-scripts.html

File: docs/programming-guide.md [hunk 3]
```diff
@@ -743,6 +855,30 @@ def doStuff(self, rdd):
...
+<div data-lang="r"  markdown="1">
+
+Spark's API relies heavily on passing functions in the driver program to run on the cluster.
+There are three recommended ways to do this:
+
+* [Anonymous functions](http://adv-r.had.co.nz/Functional-programming.html#anonymous-functions),
+  for simple functions that can be written as an anonymous function.
+* Top-level functions in a module.
+
+For example, to pass a longer function than can be supported using a `lambda`, consider
```
Comment (cafreeman) [inline docs/programming-guide.md]: Is this a carryover from the PySpark docs? Probably better not to use `lambda` in the R version.

File: docs/programming-guide.md [hunk 4]
```diff
@@ -743,6 +855,30 @@ def doStuff(self, rdd):
...
+the code below:
+
+{% highlight r %}
+"""MyScript.py"""
+myFunc <- funciton(s) {
```
Comment (cafreeman) [inline docs/programming-guide.md]: nit: `function` is misspelled.

File: docs/programming-guide.md [hunk 5]
```diff
@@ -907,6 +1055,30 @@ We could also use `counts.sortByKey()`, for example, to sort the pairs alphabeti
...
+many times each line of text occurs in a file:
+
+{% highlight r %}
+lines <- textFile(sc, "data.txt")
+pairs <- map(lines, function(s) list(s, 1))
+counts <- reduceByKey(pairs, function(a, b){a + b})
+{% endhighlight %}
+
+We could also use `counts.sortByKey()`, for example, to sort the pairs alphabetically, and finally
+`counts.collect()` to bring them back to the driver program as a list of objects.
```
Comment (cafreeman) [inline docs/programming-guide.md]: Both code snippets should be changed to the R syntax, e.g `collect(counts)`.

File: docs/quick-start.md [hunk 6]
```diff
@@ -214,6 +286,24 @@ tens or hundreds of nodes. You can also do this interactively by connecting `bin
...
+<div data-lang="r" markdown="1">
+
+{% highlight r %}
+> cache(linesWithSpark)
+
+> count(linesWithSpark)
+[1] 15
+
+> count(linesWithSpark)
+[1] 15
```
Comment (cafreeman) [inline docs/quick-start.md]: Is `count` supposed to be repeated?

**Comment by cafreeman** [comment]
Left some comments inline, but most of them seem like minor details leftover from translating the `PySpark` docs. Overall I think this is looking really good.

File: docs/index.md [hunk 7]
```diff
@@ -54,6 +54,15 @@ Example applications are also provided in Python. For example,
 
     ./bin/spark-submit examples/src/main/python/pi.py 10
 
+Spark also provides a R API. To run Spark interactively in a R interpreter, use
```
Comment (shivaram) [inline docs/index.md]: I think here (or somewhere close by) we should say that SparkR is an experimental component in `<SPARK_VERSION>` and that only the RDD API and DataFrame APIs have been implemented in SparkR.

File: docs/programming-guide.md [hunk 8]
```diff
@@ -1582,4 +1771,4 @@ For help on deploying, the [cluster mode overview](cluster-overview.html) descri
 in distributed operation and supported cluster managers.
 
 Finally, full API documentation is available in
-[Scala](api/scala/#org.apache.spark.package), [Java](api/java/) and [Python](api/python/).
+[Scala](api/scala/#org.apache.spark.package), [Java](api/java/), [Python](api/python/) [R](api/R/).
```
Comment (shivaram) [inline docs/programming-guide.md]: Minor nit: Comma after python here

File: docs/quick-start.md (hunk 6 above)
Comment (davies) [inline docs/quick-start.md]: yes

File: docs/programming-guide.md (hunk 2 above)
Comment (davies) [inline docs/programming-guide.md]: The EC2 link applies to all the languages, so I'd like to leave out of this section.

File: R/pkg/R/pairRDD.R [hunk 9]
```diff
@@ -327,7 +327,7 @@ setMethod("reduceByKey",
               convertEnvsToList(keys, vals)
             }
             locallyReduced <- lapplyPartition(x, reduceVals)
-            shuffled <- partitionBy(locallyReduced, numPartitions)
+            shuffled <- partitionBy(locallyReduced, as.integer(numPartitions))
```
Comment (shivaram) [inline R/pkg/R/pairRDD.R]: We should use `numToInt` from utils.R here

File: docs/index.md [hunk 10]
```diff
@@ -54,6 +54,15 @@ Example applications are also provided in Python. For example,
 
     ./bin/spark-submit examples/src/main/python/pi.py 10
 
+Spark also provides an experimental R API since 1.4 (only RDD and DtaFrame APIs included).
```
Comment (shivaram) [inline docs/index.md]: DataFrames spelling typo

File: docs/quick-start.md [hunk 11]
```diff
@@ -171,6 +205,44 @@ Here, we combined the [`flatMap`](programming-guide.html#transformations), [`map
...
+{% highlight r %}
+> mymax <- function(a, b) { max(a, b) }
+> reduce(map(textFile, function(line) { length(strsplit(line, " ")[[1]]) }), mymax)
+[1] 14
+{% endhighlight %}
+
+One common data flow pattern is MapReduce, as popularized by Hadoop. Spark can implement MapReduce flows easily:
+
+{% highlight r %}
+> wordCounts <- reduceByKey(map(flatMap(textFile, function(line) strsplit(line, " ")[[1]]), function(word) list(word, 1)), "+", 2)
```
Comment (shivaram) [inline docs/quick-start.md]: Can you split this into multiple lines, so it becomes more readable ? i.e. something like
```
wordCounts <- reduceByKey(
  map(
    flatMap(textFile, function(line) strsplit(line, " ")[[1]]), 
    function(word) list(word, 1)), 
 "+",  2)
```

File: docs/quick-start.md [hunk 12]
```diff
@@ -214,6 +286,24 @@ tens or hundreds of nodes. You can also do this interactively by connecting `bin
 a cluster, as described in the [programming guide](programming-guide.html#initializing-spark).
 
 </div>
+<div data-lang="r" markdown="1">
+
+{% highlight r %}
+> cache(linesWithSpark)
+
+> count(linesWithSpark)
```
Comment (shivaram) [inline docs/quick-start.md]: I get 19 here ?

File: docs/programming-guide.md [hunk 13]
```diff
@@ -576,6 +660,34 @@ before the `reduce`, which would cause `lineLengths` to be saved in memory after
...
+The second line defines `lineLengths` as the result of a `map` transformation. Again, `lineLengths`
+is *not* immediately computed, due to laziness.
+Finally, we run `reduce`, which is an action. At this point Spark breaks the computation into tasks
+to run on separate machines, and each machine runs both its part of the map and a local reduction,
+returning only its answer to the driver program.
+
+If we also wanted to use `lineLengths` again later, we could add:
+
+{% highlight r %}
+persist(lineLengths)
```
Comment (shivaram) [inline docs/programming-guide.md]: This should either be `persist(lineLengths, "MEMORY_ONLY")` or `cache(lineLengths)`. The current one gives an error.

File: docs/programming-guide.md [hunk 14]
```diff
@@ -743,6 +855,29 @@ def doStuff(self, rdd):
...
+There are three recommended ways to do this:
+
+* [Anonymous functions](http://adv-r.had.co.nz/Functional-programming.html#anonymous-functions),
+  for simple functions that can be written as an anonymous function.
+* Top-level functions in a module.
+
+For example, to pass a longer function, consider the code below:
+
+{% highlight r %}
+"""MyScript.py"""
```
Comment (shivaram) [inline docs/programming-guide.md]: MyScript.R ?

**Comment by shivaram** [comment]
I did one more pass by trying out all the code samples and left a few comments. One last thing I had was that we could probably add a note in the accumulators section that they are not supported in R yet with a link to https://issues.apache.org/jira/browse/SPARK-6815

File: docs/programming-guide.md (hunk 13 above)
Comment (davies) [inline docs/programming-guide.md]: Add "MEMORY_ONLY" as default value of newLevel, to be consistent with other API.

File: docs/quick-start.md (hunk 12 above)
Comment (davies) [inline docs/quick-start.md]: This number may be outdated (The readme could be updated without touching these examples).

**Comment by davies** [comment]
@shivaram I had addressed your comments, could you take another pass?

**Comment by davies** [comment]
@shivaram Should we merge this or wait for API audit?

File: docs/programming-guide.md (hunk 13 above)
Comment (davies) [inline docs/programming-guide.md]: Added a default value for `newLevel` of `persist`

**Comment by davies** [comment]
@shivaram I had removed the R docs for RDD API, could you take another round?

File: docs/sql-programming-guide.md [hunk 15]
```diff
@@ -491,6 +573,37 @@ for teenName in teenNames.collect():
 
 </div>
 
+<div data-lang="r"  markdown="1">
+
```
Comment (shivaram) [inline docs/sql-programming-guide.md]: This is not applicable right now

Comment (davies) [inline docs/sql-programming-guide.md]: done

**Comment by davies** [comment]
@shivaram Is this ready to go?

File: docs/sql-programming-guide.md [hunk 16]
```diff
@@ -1430,6 +1633,16 @@ df = sqlContext.load(source="jdbc", url="jdbc:postgresql:dbserver", dbtable="sch
...
+{% highlight r %}
+
+df <- laodDF(sqlContext, source="jdbc", url="jdbc:postgresql:dbserver", dbtable="schema.tablename")
```
Comment (shivaram) [inline docs/sql-programming-guide.md]: Minor typo: This should be `loadDF`

**Comment by shivaram** [comment]
@davies Sorry for the delay in looking at this. I think this change looks pretty good -- I found a minor typo that we can fix up during merge. 

I think it might be better to actually create a new page for SparkR rather than append it to the DataFrames page -- but I'l do this in a follow PR.

LGTM

**Pull request closed without merging**